### PR TITLE
Profile: Use two null block terminators

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes current failures on 32-bit linux.

It seems CI was lucky in passing in https://github.com/JuliaLang/julia/pull/41742
There was a couple more places that block end detection needed the new heuristic to ignore the rogue 0's on 32-bit linux

I also made block end detection a function given it's used in various places now

### CODE REVIEW

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -276,7 +276,7 @@ end
 function get_thread_ids(data::Vector{<:Unsigned}, taskid = nothing)
     threadids = Int[]
     for i in length(data):-1:1
-        if data[i] == 0 # find start of block
+        if i > 1 && is_block_end(data, i)
```
Comment (IanButterworth) [inline stdlib/Profile/src/Profile.jl:279]: This change and the change above are the actual fix. 
The rest is just removing code duplication

**Review by vchuravy — APPROVED** [review]

**Comment by vtjnash** [comment]
Would it help to separate them by 2 NULLs? Unlike a single NULL (which can appear in a trace), a double-NULL should not be possible. Or use `-1` to separate them (which we already use to mark the start of special frames)?

**Comment by IanButterworth** [comment]
Impossible is definitely nicer, but that would likely break downstream consumers like vscode's profiler etc.
Currently it's highly unlikely that this approach wouldn't work, I believe. Perhaps that's ok?

**Comment by vtjnash** [comment]
We used to delineate with double-NULL until quite recently, in many cases, due to a bug, so that would be no worse than before :joy:.

I was also assuming this function may still normalize to alter the marker data, as the PR does now, IIUC.

**Comment by IanButterworth** [comment]
Sounds good. Just implemented that

**Comment by IanButterworth** [comment]
I believe this is as discussed, so I'll merge to get CI back on track

**Pull request merged**